bot/exts/filtering/_filter_lists/extension.py

```diff
@@ -116,3 +127,80 @@ async def actions_for(
         ctx.blocked_exts |= set(not_allowed)
         actions = self[ListType.ALLOW].defaults.actions if ctx.event != Event.SNEKBOX else None
         return actions, [f"`{ext}`" if ext else "`No Extension`" for ext in not_allowed], {ListType.ALLOW: triggered}
+
+
+class EmbedFileHandler(commands.Cog):
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @staticmethod
+    async def _convert_attachment(attachment: discord.Attachment) -> paste_service.PasteFile:
+        encoding = re.search(r"charset=(\S+)", attachment.content_type).group(1)
+        file_content = (await attachment.read()).decode(encoding)
+        return paste_service.PasteFile(content=file_content, name=attachment.filename)
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message) -> None:
+        # Check if the message contains an embedded file and is not sent by a bot
+        if message.author.bot or not message.attachments:
+            return
+
+        bot_reply = await message.reply(f"React with {PASTEBIN_UPLOAD_EMOJI} to upload your file to our paste bin")
+        await bot_reply.add_reaction(PASTEBIN_UPLOAD_EMOJI)
+
+        def wait_for_upload_permission(reaction: discord.Reaction, user: discord.User) -> bool:
+            return (
+                reaction.message.id == bot_reply.id
+                and str(reaction.emoji) == PASTEBIN_UPLOAD_EMOJI
+                and user == message.author
+            )
```

Comment:
wait, why are we defining this function inside the coroutine? Is that necessary? It strikes me as seems needlessly nested / convoluted, I'd just make this a helper function and give it a leading underscore, make it part of the cog class, and use ?`self._wait_for_upload_permission` below as the callback.

Reply:
I'll have to see if that's possible, since it's reading `bot_reply` and `message` from the outer scope, and I think the signature of `wait_for_upload_permission` is mandated by `self.bot.wait_for("reaction_add"`.

Reply:
hm, okay I see. yeah that's a good reason, then :)